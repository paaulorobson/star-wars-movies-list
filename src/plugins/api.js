import axios from "axios"

export default {
  install (Vue, apis) {
    apis.forEach(api => {
      const { name, apiKey } = api
      const axiosInstance = axios.create({
        baseURL: `https://${process.env[`VUE_APP_${name.toUpperCase()}_API`]}`
      })

      if (apiKey) {
        axiosInstance.interceptors.request.use((config) => {
          config.params = config.params || {}
          config.params.api_key = process.env[`VUE_APP_${name.toUpperCase()}_KEY`]
          return config
        })
      }

      Vue.prototype[`$${name}`] = axiosInstance
    })
  }
}
