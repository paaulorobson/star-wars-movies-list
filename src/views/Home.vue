<template>
    <loading-indicator v-if="!isLoaded" />
    <div class="home" v-else>
        <section class="films">
            <ul class="films__timeline flex flex-col w-full m-auto">
                <li
                    v-for="film in sortedFilmsChronologically"
                    :key="film.episode_id"
                    @click="viewDetails(film.episode_id)"
                    class="film heading w-1/2 my-32"
                >
                    <h2 class="film__title">
                        {{ film.title.toUpperCase() }}
                    </h2>
                    <span class="film__release text-xl text-white">{{ film.release_date | moment("Y") }}</span>
                    <span class="film__episode serif">{{ film.episode_id | toRomanNumeral}}</span>
                </li>
            </ul>
            <div class="films__line" />
        </section>
    </div>
</template>

<script>
import LoadingIndicator from "../components/LoadingIndicator.vue"

export default {
  components: { LoadingIndicator },
  filters: {
    toRomanNumeral (value) {
      const numbers = new Map([
        [1, "I"],
        [2, "II"],
        [3, "III"],
        [4, "IV"],
        [5, "V"],
        [6, "VI"],
        [7, "VII"],
        [8, "VIII"],
        [9, "IX"]
      ])

      return numbers.get(value)
    }
  },
  data () {
    return {
      films: [],
      isLoaded: false
    }
  },
  computed: {
    sortedFilmsChronologically () {
      const films = [...this.films]
      return films.sort((a, b) =>
        new Date(a.release_date) > new Date(b.release_date) ? 1 : -1
      )
    }
  },
  created () {
    this.fetchFilms()
  },
  methods: {
    fetchFilms () {
      this.isLoaded = false
      this.$swapi
        .get("films")
        .then(({ data }) => {
          this.films = data.results
        })
        .finally(() => {
          this.isLoaded = true
        })
    },
    viewDetails (id) {
      if (id) {
        this.$router.push({
          name: "Film",
          params: {
            episodeId: id,
            movie: this.films.find(film => film.episode_id === +id),
            sequel: this.films.find(film => film.episode_id === id + 1)
          }
        })
      }
    }
  }
}
</script>

<style lang="postcss">
.films {
  position: relative;

    &__line {
        height: 100%;
        width: 2px;
        background: linear-gradient(45deg, #333, #eee, #333);
        border-radius: 50%;
        -webkit-transform: rotateX(5deg) translateX(-50%);
        transform: rotateX(5deg) translateX(-50%);
        top: 0;
        position: absolute;
        left: 50%;
        box-shadow: 0 0 10px rgba(255, 255, 255, 1), 0 0 20px rgba(255, 255, 255, 1);
    }

    &__timeline {
        @apply overflow-hidden;
        .film {
            cursor: pointer;
            @apply relative px-6;

            > * {
                z-index: 1;
                position: relative;
            }

            &__title {
                font-size: 3vw;
                transition: 100ms color ease-in;
                @apply text-yellow-500 leading-tight;
            }

            &__episode {
              font-weight: bolder;
              color: rgba(0,0,0,.95);
              text-shadow: 0 0 10px rgba(255, 255, 255, .5);
              font-size: 15vw;
              line-height: 0;
              top: 50%;
              left: 2rem;
              transform: translateY(-50%);
              z-index: 0;
              transition: 100ms color ease-in;
              @apply absolute;
            }

            &:nth-child(odd) {
                @apply pr-24 text-right;
                transform: perspective(100px) rotateX(5deg) rotateY(2deg);

                .film__episode {
                    right: 2rem;
                }
            }

            &:nth-child(even) {
                @apply pl-24 ml-auto;
                transform: perspective(100px) rotateX(5deg) rotateY(-2deg);
            }

            &:hover {
              .film__title {
                color: #FFF;
              }

              .film__episode {
                @apply text-yellow-500;
              }
            }
        }
    }
}
</style>
