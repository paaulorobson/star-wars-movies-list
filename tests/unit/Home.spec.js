import { createLocalVue, shallowMount } from "@vue/test-utils"
import Home from "@/views/Home.vue"
import api from "@/plugins/api"

const localVue = createLocalVue()

/* Load default plugins */
localVue.use(api, [
  { name: "swapi" },
  { name: "tmdb", apiKey: true }
])
localVue.use(require("vue-moment"))

/* Set wrapper */
const wrapper = shallowMount(Home, {
  localVue
})

/* Mocks data */
wrapper.setData({
  films: [{ title: "A New Hope", episode_id: 4, opening_crawl: "It is a period of civil war.\r\nRebel spaceships, striking\r\nfrom a hidden base, have won\r\ntheir first victory against\r\nthe evil Galactic Empire.\r\n\r\nDuring the battle, Rebel\r\nspies managed to steal secret\r\nplans to the Empire's\r\nultimate weapon, the DEATH\r\nSTAR, an armored space\r\nstation with enough power\r\nto destroy an entire planet.\r\n\r\nPursued by the Empire's\r\nsinister agents, Princess\r\nLeia races home aboard her\r\nstarship, custodian of the\r\nstolen plans that can save her\r\npeople and restore\r\nfreedom to the galaxy....", director: "George Lucas", producer: "Gary Kurtz, Rick McCallum", release_date: "1977-05-25" }],
  isLoaded: true
})

test("Home matches the snapshot", () => {
  expect(wrapper.element).toMatchSnapshot()
})

test("A title calls viewDetails when clicked", () => {
  const viewDetailsStub = jest.fn()
  wrapper.setMethods({ viewDetails: viewDetailsStub })
  wrapper.find("li.film").trigger("click")
  expect(viewDetailsStub).toHaveBeenCalled()
})
