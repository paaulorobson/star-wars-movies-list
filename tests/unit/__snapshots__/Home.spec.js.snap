// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home matches the snapshot 1`] = `
<div
  class="home"
>
  <section
    class="films"
  >
    <ul
      class="films__timeline flex flex-col w-full m-auto"
    >
      <li
        class="film heading w-1/2 my-32"
      >
        <h2
          class="film__title"
        >
          
                    A NEW HOPE
                
        </h2>
         
        <span
          class="film__release text-xl text-white"
        >
          1977
        </span>
         
        <span
          class="film__episode serif"
        >
          IV
        </span>
      </li>
    </ul>
     
    <div
      class="films__line"
    />
  </section>
</div>
`;
